articles/understanding-differential-privacy/index.md

```diff
@@ -0,0 +1,297 @@
+ tensor([ True,  True,  True, False,  True,  True,  True,  True,  True, False,
+         False,  True, False, False,  True,  True, False, False, False]),
+ tensor([ True,  True,  True, False,  True,  True,  True,  True,  True, False,
+         False,  True, False, False,  True,  True, False, False, False]), .................
+ tensor([ True,  True,  True, False,  True,  True,  True,  True,  True, False,
+         False,  True, False, False,  True,  True, False, False, False]),
+```
+
+Now, we create a set of databases based on the users' input
+
+```python
+  def create_db_and_parallels(num_entries):
+      db = torch.rand(num_entries) > 0.5
+      pdbs = get_parallel_dbs(db)
+      return db, pdbs
+
+  db, pdbs = create_db_and_parallels(5000) # Create 5000 different parallel databases with each database having 1 missing value
+```
+
+#### Evaluating differential privacy of a function
+
+Having created a sample database for demonstration of Differential privacy. We should be able to query the information and check if the query leaks private information or not.
+
+Evaluation of privacy leakage is measured in terms of `sensitivity`. On iterating through each row of the database, we measure the difference in the output of the query.
+
+Finding the sum of all values in the Original database (without removing values)
+
+```python
+  # A query to find the sum of values in each of the databases
+  def query(db):
+      return db.sum()
+
+  full_db_result = query(db) # Store the summation
+  print(full_db_result)
+```
+
+**Output:**
+
+```bash
+Out[4]: tensor(2454)
+```
+
+Finding the sum of all values in New database (each containing 1 missing value)
+
+```python
+  sensitivity = 0 # Assume there is no leakage
+  for pdb in pdbs:
+      pdb_result = query(pdb) # Query each new database
+
+      db_distance = torch.abs(pdb_result - full_db_result) # Compare the new database with original database
+
+      if(db_distance > sensitivity):
+          sensitivity = db_distance # Measure if privacy has been leaked
+  print(sensitivity)
+```
+
+**Output:**
+
+```bash
+>> tensor(1)
+```
+
+This demonstrates that the new set of databases containing 1 missing value in each of the databases has leaked information about the missing value. Thus, we conclude that the privacy of that person has been leaked.
+
+Now, let us have a look at how the leakage can be resolved.
+
+### Types of Differential privacy
+
+Database administrators have all rights to query anything from the database, but what if we don't give our original data in the database. Even if they query for finding leakage, they won't be able to find if the data is valid or not. That leads to the next idea of adding noise to the inputs.
+
+#### 1) Local Differential Privacy
+
+- Adding noise to the function inputs. This is done by the users, who share their data and add to the database. It is more secure, in case the Database Administrators is not trustworthy.
```

Comment:
_This is done by the users,_

Is noise addition added by the users, or the database curators?